### DISCUSSION TITLE
Is there any documentation on how paratest works? I tried to look for it but without any luck :/

### DISCUSSION THREAD

**Discussion by hostaway-greg** [paratestphp/paratest#846] [General]
Hi,

I have few questions on how paratest works, but I cannot find any answer :/ 
How does test selection work? 
1. Is it scanning a directory for tests, then divides them into groups and assign a worker to a group of tests?
2. Is it scanning a directory for tests, spawning workers, and picks a test out of the discovered tests?
3. other? 

Can it use result cache to order test by execution time?
If yes, then does it affect how test selection/worker assignment works from the previous question? 

Does small/medium/large annotation/attribute affect test selection/assignment to a worker? 

Are there any entry points to modify default behavior ? 

Thank you and have a nice day

**Comment by Slamdunk** [comment]
> How does test selection work?

1. ParaTest asks PHPUnit in [`SuiteLoader`](https://github.com/paratestphp/paratest/blob/7.x/src/WrapperRunner/SuiteLoader.php) to get a list of all the `PHPUnit\Framework\TestCase` and `PHPUnit\Runner\PhptTestCase`
2. [`WrapperRunner` gets that list](https://github.com/paratestphp/paratest/blob/ae5cc2443423ab6f4e23af0868c2e343438f3274/src/WrapperRunner/WrapperRunner.php#L112)
3. [`WrapperRunner` starts _n_ empty workes](https://github.com/paratestphp/paratest/blob/ae5cc2443423ab6f4e23af0868c2e343438f3274/src/WrapperRunner/WrapperRunner.php#L115)
4. As soon as a worker gets free, it receivse the next test to run, see [`WrapperRunner:assignAllPendingTests`](https://github.com/paratestphp/paratest/blob/7.x/src/WrapperRunner/WrapperRunner.php#L129)

> Can it use result cache to order test by execution time?

[Not yet](https://github.com/paratestphp/paratest/blob/ae5cc2443423ab6f4e23af0868c2e343438f3274/src/WrapperRunner/SuiteLoader.php#L66) because while reading the cache is easy, writing it isn't. Any PR is welcome.

> Does small/medium/large annotation/attribute affect test selection/assignment to a worker?

Test selection is the same as PHPUnit's one.
Test assignement isn't affected by test selection.

> Are there any entry points to modify default behavior ?

No, and that's by design: reliable parallelization is a nightmare, and I don't want issues opened for unknown/external code behaviour

**Reply by hostaway-greg** [reply]
Great :) Thanks for this information. Maybe it should be added to a README? 
I will look into that cache, maybe I could help. But no promisses.

**Comment by uncaught** [comment]
I would like to know some of the flow, too. Especially those things:

- How does paratest run class level code (`setUpBeforeClass`/`tearDownAfterClass`)?
    - Do those always run for every test case individually (process isolated)
    - or just by chance depending on whether the tests were assigned to the same process
    - or are test cases in one class always executed in one process?
- How does paratest handle dependent test cases (`@depends`/`#[Depends()]`)?

Thank you

**Reply by mastacheata** [reply]
By default Paratest splits your tests by TestCase across the available processes.
If you run it in functional mode where tests can be parallelised as well, I would assume the setUpBeforeClass is not offering any benefits.
